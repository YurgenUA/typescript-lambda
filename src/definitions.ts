import { AccommodationSaleableUnit } from './model/generated/AccommodationSaleableUnit';
import { AccommodationSaleableComponent } from './model/generated/AccommodationSaleableComponent';

export default class DocDbSecurityCredentials {
  username: string;
  password: string;
  engine: string;
  host: string;
  port: string;
  ssl: boolean;
  dbClusterIdentifier: string;
}

export interface OutgoingMessageData {
  component: AccommodationSaleableComponent;
  units: AccommodationSaleableUnit[];
}

export interface OutgoingMessage {
  data: OutgoingMessageData;
  metadata: {
    correlationId: string;
  }
}
