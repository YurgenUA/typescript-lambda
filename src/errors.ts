export class MongoDuplicatesError extends Error {
  constructor() {
    super('Duplicates while storing to database.');
  }
}
