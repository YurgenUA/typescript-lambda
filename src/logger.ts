import * as winston from 'winston';
import { localStorage } from './utils';

const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || 'info',
  format: winston.format.combine(
    winston.format.errors({ stack: true }),
    winston.format.json(),
  ),
  defaultMeta: {
    logger_name: 'product-accommodation-saleable',
    get timestamp() {
      return new Date().toISOString();
    },
    get correlationId() {
      return localStorage.getStore();
    }
  },

  transports: [
    new winston.transports.Console({
      format: winston.format.simple(),
    })],
});

export default logger;
