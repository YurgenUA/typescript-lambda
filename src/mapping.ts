import { v4 as uuidv4 } from 'uuid';
import * as _ from 'lodash';
import AccommodationInputModel from './model/accommodationInputModel';
import { AccommodationSaleableComponent } from './model/generated/AccommodationSaleableComponent';
import { AccommodationSaleableUnit } from './model/generated/AccommodationSaleableUnit';
import AccommodationGeographyModel, { ServedBy } from './model/accommodationGeographyModel';
import { omitByDeep } from './utils';
import { AccommodationBaselineUnit } from './model/generated/AccommodationBaselineUnit';

const COMPONENT_SCHEMA_VERSION = '0.2.6';
const UNIT_SCHEMA_VERSION = '0.2.4';

function getServedBy(servedBy: ServedBy) {
  return {
    transportHubId: servedBy.transportHubId,
    transportHubCode: servedBy.transportHubCode,
    transportHubType: servedBy.servedByType,
    sourceMarket: servedBy.sourceMarket,
    validDateBand: {
      startDate: servedBy?.validDateBand?.fromDate,
      endDate: servedBy?.validDateBand?.toDate
    }
  };
}

export function createGeographies(inputGeography: AccommodationGeographyModel): any {
  return {
    logicalLocations: inputGeography.logicalLocations && inputGeography.logicalLocations.length !== 0 ? inputGeography.logicalLocations : undefined,
    structuralLocations: [{
      country: inputGeography.country,
      destination: inputGeography.destination,
      resort: inputGeography.resort
    }],
    isServedBy: inputGeography.isServedBy?.map((served) => getServedBy(served))
  };
}

export function createSaleableComponent(input: AccommodationInputModel): AccommodationSaleableComponent {
  const { component } = _.cloneDeep(input);

  delete component.baselineComponentVersion; // eslint-disable-line @typescript-eslint/dot-notation
  delete component.sourcedComponentData.version; // eslint-disable-line @typescript-eslint/dot-notation
  component.accommodationComponent?.attributes?.forEach((e) => {
    delete e.subType;
    delete e.description;
  });

  let inactiveReason = null;
  if (!input.geographies) {
    inactiveReason = ['NO_GEO'];
  } else if (input.geographies.length > 1) {
    inactiveReason = ['MULTIPLE_GEO'];
  }

  return omitByDeep({
    ...component,
    schemaVersion: COMPONENT_SCHEMA_VERSION,
    saleableComponentId: uuidv4(),
    saleableComponentVersion: 1,
    entityStatus: input.geographies && input.geographies.length === 1 ? 'ACTIVE' : 'INACTIVE',
    inactiveReason,
    username: 'Automatic',
    creationTimeStamp: new Date(), // overwriting date
    updatedTimeStamp: new Date(),
    accommodationComponent: {
      sourceMarketClassifications: component.accommodationComponent.officialClassifications,
      geographies: input.geographies && input.geographies.length === 1 ? createGeographies(input.geographies[0]) : undefined,
      ...component.accommodationComponent,
    },
  });
}

export function createSaleableUnit(input: AccommodationBaselineUnit,
  geographies: Array<AccommodationGeographyModel>,
  componentId: string): AccommodationSaleableUnit {
  const unit = _.cloneDeep(input);

  delete unit.baselineComponentId;
  delete unit.baselineUnitVersion;
  delete unit.termsAndConditions;
  delete unit.sourcedUnitData.roomTypeVersion; // eslint-disable-line @typescript-eslint/dot-notation
  if (unit?.accommodationUnit) {
    unit.accommodationUnit.name = unit?.sourcedUnitData?.roomTypeName;
  }

  return omitByDeep({
    ...unit,
    schemaVersion: UNIT_SCHEMA_VERSION,
    saleableUnitId: uuidv4(),
    saleableUnitVersion: 1,
    entityStatus: getUnitStatus(unit, geographies),
    saleableComponentId: componentId,
    creationTimeStamp: new Date(), // overwriting date
    updatedTimeStamp: new Date(),
    default: true,
  });
}

export function getUnitStatus(unit: AccommodationSaleableUnit, geographies: Array<AccommodationGeographyModel>) {
  return geographies && geographies.length === 1 ? 'ACTIVE' : 'INACTIVE';
}
