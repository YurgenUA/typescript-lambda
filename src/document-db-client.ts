import * as awsXRay from 'aws-xray-sdk-core';
import * as awssdk from 'aws-sdk';
import { MongoClient } from 'mongodb';
import * as fs from 'fs';
import * as path from 'path';
import logger from './logger';
import DocDbSecurityCredentials from './definitions';
import { MongoDuplicatesError } from './errors';

const { AWS_STAGE, DB_NAME, TIMEOUT } = process.env;
const AwsSdk = awsXRay.captureAWS(awssdk);
const { SecretsManager } = AwsSdk;

const ca = [fs.readFileSync(path.join(__dirname, '../rds-combined-ca-bundle.pem'))];
const secretsManager = new SecretsManager();
const COLLECTIONS = {
  SALEABLE_COMPONENT: 'accommodation-saleable-components',
  SALEABLE_UNIT: 'accommodation-saleable-units',
};

export default class DocumentDbClient {
  private client: MongoClient;

  async getClient() {
    if (this.client) return this.client;
    try {
      const connString = await DocumentDbClient.getConnectionString();
      const timeoutMS = Number(TIMEOUT) * 1000;
      this.client = await MongoClient.connect(connString, {
        sslValidate: true,
        sslCA: ca,
        useNewUrlParser: true,
        useUnifiedTopology: true,
        poolSize: 1,
        connectTimeoutMS: timeoutMS,
        socketTimeoutMS: timeoutMS
      });
      return this.client;
    } catch (e) {
      logger.error(e);
      throw new Error('Failed to connect to db');
    }
  }

  static async getConnectionString(): Promise<string> {
    if (process.env.IS_LOCAL === 'true') {
      return process.env.MONGO_CONNECTION_STRING;
    }
    const secretName = `/product/${AWS_STAGE}/docdb-credentials`;
    const response = await secretsManager.getSecretValue({ SecretId: secretName }).promise();

    if (!('SecretString' in response)) {
      throw new Error('Failed to get DB credentials');
    }
    const creds: DocDbSecurityCredentials = JSON.parse(response.SecretString);
    return `mongodb://${creds.username}:${creds.password}@${creds.host}:${creds.port}/?ssl=${creds.ssl}&readPreference=secondaryPreferred&retryWrites=false`;
  }

  async saveComponentIfNeeded(record) {
    const documentDbClient = await this.getClient();
    const db = documentDbClient.db(DB_NAME);
    const collection = db.collection(COLLECTIONS.SALEABLE_COMPONENT);
    await collection.createIndex({ 'sourcedComponentData.id': 1 }, { unique: true });
    await collection.createIndex({ baselineComponentId: 1 }, { unique: true });

    logger.info(`Creating new saleable component with id: ${record.sourcedComponentData.id}`);
    try {
      return await collection.insertOne(record);
    } catch (error) {
      // if it's mongo error 11k: duplicate key error saleable component already exists for given baseline component
      if (typeof error === 'object' && error.name === 'MongoError' && error.code === 11000) {
        logger.error('Skip creating SaleableComponent because of duplication.');
        logger.error(`BaselineComponentId: ${record.baselineComponentId}. SourcedComponentId:${record.sourcedComponentData.id}`);
        throw new MongoDuplicatesError();
      }
      throw error;
    }
  }

  async updateComponent(record) {
    const documentDbClient = await this.getClient();
    const db = documentDbClient.db(DB_NAME);
    const collection = db.collection(COLLECTIONS.SALEABLE_COMPONENT);

    logger.info(`Updating saleable component with id: ${record.sourcedComponentData.id}`);
    await collection.updateOne({ 'sourcedComponentData.id': record.sourcedComponentData.id }, { $set: record });
    await collection.updateOne({ 'sourcedComponentData.id': record.sourcedComponentData.id }, { $unset: { inactiveReason: 1 } });
  }

  async saveUnitIfNeeded(record) {
    const documentDbClient = await this.getClient();
    const db = documentDbClient.db(DB_NAME);
    const collection = db.collection(COLLECTIONS.SALEABLE_UNIT);
    await collection.createIndex({ 'sourcedUnitData.roomTypeId': 1 }, { unique: true });
    await collection.createIndex({ baselineUnitId: 1 }, { unique: true });

    logger.info(`Creating new saleable unit with room type id: ${record.sourcedUnitData.roomTypeId}`);
    try {
      return await collection.insertOne(record);
    } catch (error) {
      // if it's mongo error 11k: duplicate key error saleable component already exists for given baseline component
      if (typeof error === 'object' && error.name === 'MongoError' && error.code === 11000) {
        logger.error('Skip creating SaleableUnit because of duplication.');
        logger.error(`BaselineUnitId: ${record.baselineUnitId}. RoomTypeId:${record.sourcedUnitData.roomTypeId}`);
        throw new MongoDuplicatesError();
      }
      throw error;
    }
  }

  async updateUnit(record) {
    const documentDbClient = await this.getClient();
    const db = documentDbClient.db(DB_NAME);
    const collection = db.collection(COLLECTIONS.SALEABLE_UNIT);

    logger.info(`Updating saleable unit with id: ${record.sourcedUnitData.roomTypeId}`);

    await collection.updateOne({ 'sourcedUnitData.roomTypeId': record.sourcedUnitData.roomTypeId }, { $set: record });
  }

  async loadComponentByBaselineId(baselineId: string) {
    const documentDbClient = await this.getClient();
    const db = documentDbClient.db(DB_NAME);

    return db.collection(COLLECTIONS.SALEABLE_COMPONENT).findOne({ baselineComponentId: baselineId }, { projection: { _id: 0 } });
  }

  async loadUnitByBaselineId(baselineId: string) {
    const documentDbClient = await this.getClient();
    const db = documentDbClient.db(DB_NAME);

    return db.collection(COLLECTIONS.SALEABLE_UNIT).findOne({ baselineUnitId: baselineId }, { projection: { _id: 0 } });
  }

  async loadUnitsBySaleableComponentId(saleableComponentId: string) {
    const documentDbClient = await this.getClient();
    const db = documentDbClient.db(DB_NAME);

    return db.collection(COLLECTIONS.SALEABLE_UNIT).find({ saleableComponentId }, { projection: { _id: 0 } }).toArray();
  }
}
