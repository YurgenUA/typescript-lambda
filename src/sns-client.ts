import * as awsXRay from 'aws-xray-sdk-core';
import * as awssdk from 'aws-sdk';
import logger from './logger';

const AwsSdk = awsXRay.captureAWS(awssdk);
const { SNS } = AwsSdk;
const sns = new SNS();
export async function publishSnsTopic(message: any): Promise<void> {
  const params = {
    Message: JSON.stringify(message),
    TopicArn: process.env.OUTPUT_TOPIC,
  };

  try {
    const response = await sns.publish(params).promise();

    logger.info(`Published MessageId: ${response.MessageId}`);
  } catch (error) {
    logger.error('PublishSnsTopic failed', error);
    throw error;
  }
}
