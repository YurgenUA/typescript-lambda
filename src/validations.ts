import * as Ajv from 'ajv';
import logger from './logger';

export function validateAgainstSchema(obj: object, inputSchema: any): string {
  const ALREADY_PATCHED = 'alreadyPatched';
  const TYPE_OBJECT = 'object';
  const schema = inputSchema;
  if (schema[ALREADY_PATCHED] == null) {
    schema[ALREADY_PATCHED] = true;
    const schemaObjects = new Array<object>();
    schemaObjects.push(schema.properties);
    while (schemaObjects.length > 0) {
      const currentObject = schemaObjects.pop();
      for (const value of Object.values(currentObject)) {
        const oneProperty: any = value;
        // patch json schema as ajv validates Date instances only if it's marked as 'object' with 'date-time' format
        if (oneProperty.type === 'string' && oneProperty.format === 'date-time') {
          oneProperty.type = TYPE_OBJECT;
        } else if (oneProperty.type === TYPE_OBJECT) {
          schemaObjects.push(oneProperty.properties);
        }
      }
    }
  }
  const validate = new Ajv().compile(schema);

  if (!validate(obj)) {
    const errorMessage = validate.errors.reduce((_, cur) => `${cur.dataPath}:${cur.message}\n`, '');
    const errMessage = `Output validation error(s): ${errorMessage}`;
    logger.error(errMessage);
    return errMessage;
  }
  return null;
}
