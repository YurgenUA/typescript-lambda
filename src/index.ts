import { SQSEvent, SQSRecord } from 'aws-lambda';
import * as awsSdk from 'aws-sdk';
import * as awsXRay from 'aws-xray-sdk-core';
import { v4 as uuidv4 } from 'uuid';
import logger from './logger';
import AccommodationInputModel, { IncomingMessage } from './model/accommodationInputModel';
import { createGeographies, createSaleableComponent, createSaleableUnit } from './mapping';

import DocumentDbClient from './document-db-client';
import { validateAgainstSchema } from './validations';
import { AccommodationSaleableComponent } from './model/generated/AccommodationSaleableComponent';
import { publishSnsTopic } from './sns-client';
import { AccommodationSaleableUnit } from './model/generated/AccommodationSaleableUnit';
import { MongoDuplicatesError } from './errors';
import { formatOutputMessage, getLoggingContext, withLoggingContext } from './utils';

const componentSchema = require('../schema/AccommodationSaleableComponentSchema.json');
const unitSchema = require('../schema/AccommodationSaleableUnitSchema.json');

const { SQS } = awsXRay.captureAWS(awsSdk);
const sqs = new SQS();

const dbClient = new DocumentDbClient();

function prepareMessageForSending(component: AccommodationSaleableComponent, units: AccommodationSaleableUnit[]) {
  const saleableComponent = component;
  delete saleableComponent['_id']; // eslint-disable-line @typescript-eslint/dot-notation
  const saleableUnits: any[] = units.map((unit: any) => {
    const patchedUnit = unit;
    delete patchedUnit['_id']; // eslint-disable-line @typescript-eslint/dot-notation
    return patchedUnit;
  });

  return formatOutputMessage(getLoggingContext(), { component: saleableComponent, units: saleableUnits });
}

async function processRecord(record: SQSRecord) {
  try {
    const incomingMessage: IncomingMessage = JSON.parse(record.body);
    const correlationId = incomingMessage.metadata?.correlationId || uuidv4();
    return await withLoggingContext(async () => {
      try {
        await processSingleRecord(incomingMessage.data);
      } catch (error) {
        if (error instanceof MongoDuplicatesError) {
          throw error;
        }
        logger.error('Failed to process message', error);
        await sqs.sendMessage({ QueueUrl: process.env.SQS_DLQ, MessageBody: record.body }).promise();
      }
    }, correlationId);
  } catch (e) {
    logger.error(`Process failed with error ${e.message}`);
    throw e;
  }
}

async function processSingleRecord(inputData: AccommodationInputModel) {
  logger.info(`Received baseline accommodation component ${inputData.component.baselineComponentId}`);
  logger.info(`Accommodation id is ${inputData.component.sourcedComponentData?.id}`);
  logger.info(`Received ${inputData.units?.length} baseline accommodation units`);
  for (const unit of inputData.units || []) {
    logger.info(`Received baseline accommodation unit ${unit.baselineUnitId}`);
    logger.info(`RoomTypeId is ${unit.sourcedUnitData?.roomTypeId}`);
  }
  const saleableUnits = [];
  const tasks = [];

  let saleableComponent: AccommodationSaleableComponent = await dbClient.loadComponentByBaselineId(inputData.component.baselineComponentId);

  if (!saleableComponent) {
    logger.info(`Saleable component does not exist for baseline component ${inputData.component?.baselineComponentId}. Creating new.`);
    saleableComponent = createSaleableComponent(inputData);

    const validationErrors = validateAgainstSchema(saleableComponent, componentSchema);
    if (validationErrors) {
      throw Error(validationErrors);
    }

    await dbClient.saveComponentIfNeeded(saleableComponent);
    logger.info(`Stored saleable accommodation component ${saleableComponent.saleableComponentId}`);
  } else {
    logger.info(`Loaded existing saleable component ${saleableComponent.saleableComponentId}`);
    if (saleableComponent.entityStatus === 'INACTIVE' && inputData.geographies && inputData.geographies.length === 1) {
      logger.info(`Saleable component ${saleableComponent.saleableComponentId} was inactive and geographies arrived - need to update`);
      saleableComponent.entityStatus = 'ACTIVE';
      saleableComponent.saleableComponentVersion += 1;
      const geographies = createGeographies(inputData.geographies[0]);
      saleableComponent.accommodationComponent.geographies = {
        structuralLocations: geographies.structuralLocations
      };
      if (geographies.logicalLocations) {
        saleableComponent.accommodationComponent.geographies.logicalLocations = geographies.logicalLocations;
      }
      if (geographies.isServedBy) {
        saleableComponent.accommodationComponent.geographies.isServedBy = geographies.isServedBy;
      }
      await dbClient.updateComponent(saleableComponent);
      logger.info(`Updated geography on saleable component ${saleableComponent.saleableComponentId}`);

      const existingUnits: AccommodationSaleableUnit[] = await dbClient.loadUnitsBySaleableComponentId(saleableComponent.saleableComponentId);
      if (existingUnits) {
        for (const existingUnit of existingUnits) {
          existingUnit.entityStatus = 'ACTIVE';
          existingUnit.saleableUnitVersion += 1;
          tasks.push(dbClient.updateUnit(existingUnit));
          logger.info(`Updated geography on saleable unit ${existingUnit.saleableUnitId}`);
        }
      }
    }
  }

  const errors = [];
  if (inputData.units) {
    for (const input of inputData.units) {
      // eslint-disable-next-line no-await-in-loop
      let saleableUnit: AccommodationSaleableUnit = await dbClient.loadUnitByBaselineId(input.baselineUnitId);

      if (!saleableUnit) {
        saleableUnit = createSaleableUnit(input, inputData.geographies, saleableComponent.saleableComponentId);

        const validationErrors = validateAgainstSchema(saleableUnit, unitSchema);

        if (validationErrors) {
          errors.push(Error(validationErrors));
        }
        tasks.push(dbClient.saveUnitIfNeeded(saleableUnit));
        logger.info(`Created new saleable unit ${saleableUnit.saleableUnitId}`);
        logger.info(`RoomTypeId ${saleableUnit.sourcedUnitData?.roomTypeId}`);
      } else {
        logger.info(`Loaded existing saleable unit ${saleableUnit.saleableUnitId}`);
        logger.info(`RoomTypeId ${saleableUnit.sourcedUnitData?.roomTypeId}`);
      }

      saleableUnits.push(saleableUnit);
    }
  }

  if (errors.length > 0) {
    throw errors[0];
  }

  if (tasks.length === 0) {
    // do not need further processing if no units were created
    logger.info('Stop processing current event as no units were created.');
    return;
  }

  await Promise.all(tasks);

  if (saleableComponent.entityStatus === 'ACTIVE') {
    logger.info(`Sending out to SNS saleable component ${saleableComponent.saleableComponentId}`);
    for (const saleableUnit of saleableUnits) {
      logger.info(`Sending out to SNS saleable unit ${saleableUnit.saleableUnitId}`);
    }
    await publishSnsTopic(prepareMessageForSending(saleableComponent, saleableUnits));
  } else {
    logger.info(`saleable component and units will not be sent. Component is not active. Reasons: ${saleableComponent.inactiveReason}`);
  }
}

export async function handler(event: SQSEvent) {
  try {
    logger.info('event', event);
    logger.info(`Received [${event?.Records?.length}] records from SQS`);

    const tasks = event.Records.map((record) => processRecord(record));

    await Promise.all(tasks);
    return 'OK';
  } catch (error) {
    logger.error('Lambda execution failed', error);
    throw error;
  }
}

export const testables = {
  prepareMessageForSending,
  processRecord,
  dbClient,
};
