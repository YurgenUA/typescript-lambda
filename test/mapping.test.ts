import MockDate from 'mockdate';
import * as uuid from 'uuid';
import { createSaleableComponent, createSaleableUnit } from '../src/mapping';
import AccommodationInputModel from '../src/model/accommodationInputModel';
import { validateAgainstSchema } from '../src/validations';
import { AccommodationBaselineUnit } from '../src/model/generated/AccommodationBaselineUnit';

const componentSchema = require('../schema/AccommodationSaleableComponentSchema.json');

jest.mock('uuid', () => ({
  ...jest.requireActual('uuid')
}));

describe('mapping tests', () => {
  const event = require('./events/accommodation-baseline-event.json');
  let inputModel: AccommodationInputModel;
  let uuidMock = jest.spyOn(uuid, 'v4').mockReturnValue('initial-mocked-id');

  afterEach(() => {
    uuidMock.mockRestore();
    MockDate.reset();
  });

  beforeEach(() => {
    const eventJson: string = JSON.stringify(event.data);
    inputModel = JSON.parse(eventJson);
  });

  test('component should pass validation', async () => {
    uuidMock = jest.spyOn(uuid, 'v4').mockReturnValue('91363994-df46-489e-8915-e395e28283b6');
    MockDate.set('2021-02-26T10:25:11.801Z');

    const saleableComponent = createSaleableComponent(inputModel);
    const validationErrors = validateAgainstSchema(saleableComponent, componentSchema);

    expect(saleableComponent).toMatchSnapshot();
    expect(validationErrors).toBeNull();
  });

  test('component is inactive when resort not found', async () => {
    // given
    delete inputModel.geographies;
    // when
    const saleableComponent = createSaleableComponent(inputModel);
    // then
    expect(saleableComponent.entityStatus).toBe('INACTIVE');
  });

  test('createSaleableComponent has geography', async () => {
    const input = {
      component: {
        accommodationComponent: {},
        sourcedComponentData: {},
      },
      units: [],
      geographies: [{
        country: {
          id: 'countryId',
          names: [{ name: 'countryName' }],
          code: 'CC',
        },
        destination: {
          id: 'destinationId',
          names: [{ name: 'destinationName' }],
        },
        resort: {
          id: 'resortId',
          names: [{ name: 'resortName' }],
        }
      }]
    } as AccommodationInputModel;
    const output = createSaleableComponent(input);

    expect(output.accommodationComponent.geographies.structuralLocations[0]).toStrictEqual({
      country: {
        id: 'countryId',
        names: [{ name: 'countryName' }],
        code: 'CC'
      },
      destination: {
        id: 'destinationId',
        names: [{ name: 'destinationName' }]
      },
      resort: {
        id: 'resortId',
        names: [{ name: 'resortName' }]
      }
    });
    expect(output.accommodationComponent.geographies.logicalLocations).toBeUndefined();
  });

  test('createSaleableComponent overwrite creationTimeStamp', async () => {
    const newCentury = new Date(2000, 1, 1);
    const input = {
      component: {
        creationTimeStamp: newCentury.toISOString(),
        accommodationComponent: {},
        sourcedComponentData: {},
        baselineComponentId: 'any',
        baselineComponentVersion: 1
      },
      units: [],
      geographies: [{
        country: {
          id: 'countryId',
        },
        destination: {
          id: 'destinationId',
        },
        resort: {
          id: 'resortId',
        }
      }]
    } as unknown as AccommodationInputModel;
    const output = createSaleableComponent(input);

    expect(output.creationTimeStamp).not.toBe(newCentury);
  });

  test('createSaleableComponent removes null & undefined properties', async () => {
    const input = {
      component: {
        accommodationComponent: {
          chains: null
        },
        sourcedComponentData: {},
      },
      units: [],
      geographies: [{
        country: {
          id: 'countryId',
        },
        destination: {
          id: 'destinationId',
        },
        resort: {
          id: 'resortId',
        },
        logicalLocations: undefined,
      }]
    } as AccommodationInputModel;
    const output = createSaleableComponent(input);

    expect(output.accommodationComponent).not.toHaveProperty('chains');
    expect(output.accommodationComponent.geographies).not.toHaveProperty('logicalLocations');
  });

  test('createSaleableUnits has saleableComponentId', async () => {
    const input = {
      accommodationUnit: {},
      sourcedUnitData: { roomTypeName: 'Test1' }
    } as AccommodationBaselineUnit;

    const output = createSaleableUnit(input, null, '42');
    expect(output).not.toBeNull();
    expect(output.saleableComponentId).toBe('42');
  });

  test('createSaleableUnits overwrite creationTimeStamp', async () => {
    const newCentury = new Date(2000, 1, 1);
    const input = {
      creationTimeStamp: newCentury.toISOString(),
      accommodationUnit: {},
      sourcedUnitData: { roomTypeName: 'Test1' },
      baselineUnitId: 'any',
      baselineUnitVersion: 1
    } as unknown as AccommodationBaselineUnit;
    const output = createSaleableUnit(input, null, '42');

    expect(output).not.toBeNull();
    expect(output.creationTimeStamp).not.toBe(newCentury);
  });

  test('createSaleableUnits has supplier', async () => {
    const input = {
      sourcedUnitData: {
        supplier: 'supplier',
        subSupplier: 'subSupplier',
      },
      accommodationUnit: {},
    } as AccommodationBaselineUnit;

    const output = createSaleableUnit(input, null, '42');
    expect(output).not.toBeNull();
    expect(output.sourcedUnitData.supplier).toStrictEqual('supplier');
    expect(output.sourcedUnitData.subSupplier).toStrictEqual('subSupplier');
  });

  test('createSaleableUnit has accommodationUnit.name', async () => {
    const input = {
      accommodationUnit: {},
      sourcedUnitData: { roomTypeName: 'Test1' }
    } as AccommodationBaselineUnit;

    const output = createSaleableUnit(input, null, '42');
    expect(output).not.toBeNull();
    expect(output.accommodationUnit.name).toBe('Test1');
  });

  test('createSaleableUnits has supplierRoomTypeName', async () => {
    const input = {
      sourcedUnitData: {
        supplierRoomTypeName: 'Family Room'
      },
      accommodationUnit: {},
    } as AccommodationBaselineUnit;

    const output = createSaleableUnit(input, null, '42');
    expect(output).not.toBeNull();
    expect(output.sourcedUnitData.supplierRoomTypeName).toStrictEqual('Family Room');
  });
});
