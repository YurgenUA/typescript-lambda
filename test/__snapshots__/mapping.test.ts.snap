// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapping tests component should pass validation 1`] = `
Object {
  "accommodationComponent": Object {
    "componentType": Object {
      "id": "80f9c9b8-21cc-4e7f-9608-0b9bd9140f97",
      "name": "Holiday Complex",
    },
    "concepts": Array [
      Object {
        "id": "77eda6bc-f8e0-4019-bdff-c67cbea5403b",
        "name": "TUI MAGIC LIFE",
        "validDateBands": Array [
          Object {
            "endDate": "2030-01-01",
            "startDate": "2020-01-01",
          },
        ],
      },
    ],
    "coordinates": Object {
      "latitude": "27.973457",
      "longitude": "34.42388",
    },
    "geographies": Object {
      "isServedBy": Array [
        Object {
          "sourceMarket": "BE",
          "transportHubCode": "FUE",
          "transportHubId": "50e61986-6265-4e75-83ba-0391d9fcf202",
          "transportHubType": "AIRPORT",
          "validDateBand": Object {
            "endDate": "2021-03-14",
            "startDate": "2020-03-14",
          },
        },
        Object {
          "sourceMarket": "NL",
          "transportHubCode": "FUE",
          "transportHubId": "50e61986-6265-4e75-83ba-0391d9fcf202",
          "transportHubType": "AIRPORT",
          "validDateBand": Object {
            "endDate": "2023-03-14",
            "startDate": "2022-03-14",
          },
        },
        Object {
          "sourceMarket": "BE",
          "transportHubCode": "LWO",
          "transportHubId": "cc2820f6-cb89-4e55-b3a9-db4657f9ee00",
          "transportHubType": "AIRPORT",
          "validDateBand": Object {
            "endDate": "2042-01-01",
            "startDate": "2021-01-01",
          },
        },
        Object {
          "sourceMarket": "BE",
          "transportHubCode": "KBP",
          "transportHubId": "9f2b272c-f4c0-4de5-b772-2c58e00ffbe3",
          "transportHubType": "AIRPORT",
          "validDateBand": Object {
            "endDate": "2042-01-01",
            "startDate": "2021-01-01",
          },
        },
        Object {
          "sourceMarket": "BE",
          "transportHubCode": "IEV",
          "transportHubId": "574bf689-2969-4501-a69f-ba43743c7565",
          "transportHubType": "AIRPORT",
          "validDateBand": Object {
            "endDate": "2042-01-01",
            "startDate": "2021-01-01",
          },
        },
        Object {
          "sourceMarket": "BE",
          "transportHubCode": "NLV",
          "transportHubId": "23c246a1-9284-41bf-bc92-0f2113b5269c",
          "transportHubType": "AIRPORT",
          "validDateBand": Object {
            "endDate": "2042-01-01",
            "startDate": "2021-01-01",
          },
        },
      ],
      "logicalLocations": Array [
        Object {
          "id": "45a1258d-0081-4ac7-83dd-6014cdb79630",
          "locationLevel": "OVER_DESTINATIONS",
          "locationType": "REGION",
          "name": "Canary Islands",
        },
      ],
      "structuralLocations": Array [
        Object {
          "country": Object {
            "code": "ES",
            "id": "38b4459a-ae0c-4b6a-9d73-01a6ecb17780",
            "names": Array [
              Object {
                "name": "Spain",
              },
            ],
          },
          "destination": Object {
            "id": "c7d7a78b-ad4e-426e-93f8-84ec6206633f",
            "names": Array [
              Object {
                "name": "Fuerteventura",
              },
            ],
          },
          "resort": Object {
            "id": "ffa0f262-187d-4cb1-9622-4a4985068635",
            "names": Array [
              Object {
                "name": "Playa De Jandia",
              },
            ],
          },
        },
      ],
    },
    "names": Array [
      Object {
        "name": "TUI MAGIC LIFE Sharm El Sheikh",
        "validDateBand": Object {
          "endDate": "2030-01-01",
          "startDate": "2020-01-01",
        },
      },
    ],
    "officialClassifications": Array [
      Object {
        "id": "d04d55c4-822a-4e1b-952e-64189dfbd08b",
        "validDateBand": Object {
          "endDate": "2030-01-01",
          "startDate": "2020-01-01",
        },
        "value": 4,
      },
    ],
    "sourceMarketClassifications": Array [
      Object {
        "id": "d04d55c4-822a-4e1b-952e-64189dfbd08b",
        "validDateBand": Object {
          "endDate": "2030-01-01",
          "startDate": "2020-01-01",
        },
        "value": 4,
      },
    ],
  },
  "baselineComponentId": "392be304-dea0-5e6f-9257-56bff7ae2e5d",
  "creationTimeStamp": 2021-02-26T10:25:11.801Z,
  "entityStatus": "ACTIVE",
  "saleableComponentId": "91363994-df46-489e-8915-e395e28283b6",
  "saleableComponentVersion": 1,
  "schemaVersion": "0.2.6",
  "sourcedComponentData": Object {
    "code": "A0454189",
    "id": "92f1fc3f-50bc-4c71-a2a7-3a1549754781",
    "sourcingSystem": "TUIAccommodationSourcing",
  },
  "updatedTimeStamp": 2021-02-26T10:25:11.801Z,
  "username": "Automatic",
}
`;
